File: 프로그래머스/lv2/12952. N－Queen/N－Queen.py
def possible(x, y, n, col):
    for i in range(x):
        if y == col[i] or abs(y - col[i]) == x - i:
            return False
    return True

def queen(x, n, col):
    
    if x == n:
        return 1
    
    count = 0
    
    for y in range(n):
        # 다음 퀸을 놓을 수 있는 경우만 진행
        if possible(x, y, n, col):
            col[x] = y
            count += queen(x+1, n, col) 
            
    return count

def solution(n):
    
    col = [0] * n
    answer = queen(0, n, col)
    
    return answer

File: 프로그래머스/2/12952. N－Queen/N－Queen.java
class Solution {

    private int count = 0; // 가능한 배치 경우의 수

    public int solution(int n) {
        int[] columns = new int[n];
        backtrack(columns, n, 0);
        return count;
    }

    private void backtrack(int[] columns, int n, int row) {
        if (row == n) {
            count += 1;
            return;
        }

        for (int i = 0; i < n; i++) {
            if (isPossible(columns, row, i)) {
                columns[row] = i;
                backtrack(columns, n, row + 1);
            }
        }
    }

    private boolean isPossible(int[] columns, int row, int col) {
        for (int prev = 0; prev < row; prev++) {
            if (columns[prev] == col || Math.abs(columns[prev] - col) == row - prev) {
                return false;
            }
        }

        return true;
    }
}
